modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2087,6 +2088,17 @@ def concat(self, axis, others, how, sort):
             new_lengths = None
             new_widths = None
         new_partitions = self._partition_mgr_cls.concat(axis, left_parts, right_parts)
+        shape = new_partitions.shape
+        # In the case when the number of partitions significantly differs from the value
+        # specified by `NPartitions`, we can get a significant slowdown in subsequent operations.
+        if shape[0] * shape[1] > NPartitions.get() * 2:
+            df = self._partition_mgr_cls.to_pandas(new_partitions)
+            # need to recompute `new_lengths`, `new_widths`
+            (
+                new_partitions,
+                new_lengths,
+                new_widths,
+            ) = self._partition_mgr_cls.from_pandas(df, return_dims=True)
```

Comment:
I didn't found special function for repartitioning in `partition_manager.py`

Reply:
This doesn't feel correct in any of the cases. What if driver machine cannot fit the whole dataframe?.. plus it generates too much of excessive traffic.